### PR TITLE
kernel: Scheduling enhancements

### PR DESCRIPTION
**Pull request by peter-mitsis**
This set of commits contains proposed enhancements to parts of the Zephyr scheduler code to help address some of the issues in recent performance discussions.

### CODE REVIEW

File: kernel/include/kthread.h [hunk 1]
```diff
@@ -204,6 +204,13 @@ static ALWAYS_INLINE bool should_preempt(struct k_thread *thread,
 		return true;
 	}
 
+	/* Otherwise we have to be running a preemptible thread or
+	 * switching to a metairq
+	 */
+	if (thread_is_preemptible(_current) || thread_is_metairq(thread)) {
+		return true;
+	}
```
Comment (andyross) [inline kernel/include/kthread.h:212]: Worth pointing out that if we fix swap() to remove PendSV, we get this first patch's effect for free as the code you're reordering it around would disappear.

File: kernel/msg_q.c [hunk 2]
```diff
@@ -120,6 +120,14 @@ int k_msgq_cleanup(struct k_msgq *msgq)
 	return 0;
 }
 
+static inline void msgq_copy(char *dest, const char *src, size_t n)
```
Comment (andyross) [inline kernel/msg_q.c:123]: Isn't @keith-packard fixing picolibc memcpy() to be a better optimized version?  This variant might be faster in the specific benchmark with very small messages, but a bytewise copy is near-pessimal as the copies get larger (like, literally 4x slower at minimum, and more like 16x on bigcore x86/arm64 devices with wide buses and cache-aware SIMD load/store/move operations!).

Really this seems like a bad trade, and we're doing too much benchmark chasing.  If we really need to muck with this, my vote would be to "fix" the API to require alignment (via a kconfig for non-aligned compatibility or whatever) and code a word-chunked copy.

**Review by andyross — CHANGES_REQUESTED** [review]
I'm gonna be sticky and NAK that bytewise memcpy().  I believe you that it helps in this benchmark but it's really the wrong thing for anything that isn't a tiny toy message.  Can we check out the picolibc fixups to see if they work similarly?

File: kernel/include/kthread.h (hunk 1 above)
Comment (andyross) [reply kernel/include/kthread.h:212]: Oh, and I'm going to bet that the real effect is that the reordering isn't about _current at all, it's that the two cases you're moving are much more common than the SWAP_NONATOMIC edge case, so are effecting an early exit that doesn't need to check z_is_thread_timeout_active().

File: kernel/msg_q.c (hunk 2 above)
Comment (nashif) [reply kernel/msg_q.c:123]: agree with @andyross  let's not touch the libc part and focus on kernel scheduling, you can build the optimized picolibc using ```-DCONFIG_PICOLIBC_SPEED_OPTIMIZATIONS=y  -DCONFIG_PICOLIBC_USE_MODULE=y``` or you can use NEWLIB. I just did a quick test:

baseline with optimized picolibc:  

```
*** Booting Zephyr OS build v3.7.0-4722-g338d595a6a19 ***
**** Thread-Metric Message Processing Test **** Relative Time: 30
Time Period Total:  8109040
 ```
 
with your changes in this PR: 
```
*** Booting Zephyr OS build v3.7.0-4726-g3bea9b73042c ***
**** Thread-Metric Message Processing Test **** Relative Time: 30
Time Period Total:  7362848
 ```
 
Your PR without custom byte copy:
```
*** Booting Zephyr OS build v3.7.0-4727-gc8937c0a4269 ***
**** Thread-Metric Message Processing Test **** Relative Time: 30
Time Period Total:  10127756
 ```

Comment (peter-mitsis) [reply kernel/msg_q.c:123]: I have no problems dropping that change.  Incidentally, I did some additional experimentation to profile the copying performance under different copy sizes.

Copying 1 byte at a time: 7273612
Copying 2 bytes at a time: 9601154
Copying 4 bytes at a time: 11429945
Copying 8 bytes at a time: 11650579

I mention these numbers to show the limits of what could be expected from changing the copying methodology (assuming everything else stays the same).

File: kernel/include/ksched.h [hunk 3]
```diff
@@ -159,11 +165,17 @@ static inline void unpend_thread_no_timeout(struct k_thread *thread)
 	thread->base.pended_on = NULL;
 }
 
+/*
+ * In a multiprocessor system, z_unpend_first_thread() must lock the scheduler
+ * spinlock _sched_spinlock. However, in a uniprocessor system, that is not
+ * necessary as the caller has already taken precautions (almost always in the
+ * form of locking interrupts).
```
Comment (andyross) [inline kernel/include/ksched.h:172]: Can we add an __ASSERT to validate that this is true in all cases?   Basically just test that interrupts are already locked.  See the code for z_smp_cpu_mobile() for an example (though do **not** include the is_in_isr() test, as ISRs are preemptible!).

**Review by andyross — COMMENTED** [review]
Remove -1, one final nitpick.

File: kernel/msg_q.c (hunk 2 above)
Comment (andyross) [reply kernel/msg_q.c]: On arm32 the advantage of the 8 byte case is just loop unrolling, I don't think (though apply salt, not my arch) that there's a 64 bit load instruction available on non-FPU devices.

**Review by andyross — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**